src/python/pants/backend/docker/target_types.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.engine.target import COMMON_TARGET_FIELDS, Dependencies, Sources, StringField, Target
+
+
+class DockerImageSources(Sources):
+    default = ("Dockerfile",)
+    expected_num_files = 1
+    help = "Name of the Dockerfile to use when building the Docker image."
+
+
+class DockerImageVersion(StringField):
+    alias = "version"
+    default = "latest"
+    help = "Image tag to apply to built images."
+
+
+class DockerDependencies(Dependencies):
+    pass
+
+
+class DockerContextRoot(StringField):
+    alias = "context_root"
+    default = "."
```

Comment:
I'm surprised by this field: normally resources must be declared via `dependencies`. Is this simply saying how resources are loaded within the Dockerfile, whether to use a path relative to build root vs. BUILD file?

Is there a concrete use case for this now? I'm wondering if it's something that we could punt on adding until later. Even if there is a concrete use case, maybe wait to add this field until its going to be directly used?

This PR doesn't need to add every single field that will be used, it's more about getting a skeleton set up so future code review is simpler.

Reply:
This is basically the dir that Dockerfile COPY commands are relative to. When running raw docker, it makes everything under this dir available to COPY into the image. In our case it would be everything under this dir that is also a dependency, but either way you do need to specify it, we can't commit to one default, I don't think.

Reply:
As Benjy said, we want to be able to override this. But leaving it out of your BUILD file gives you a sensible default. 

Unless. We come up with a way to setup the chroot for the docker build context, so all required files are accessible for the Dockerfile instructions to use.

Reply:
That would be tricky because we can't just arbitrarily pick a directory to be the context root - we have to use the directory relative to which the COPY commands were written, which is a choice the author of the Dockerfile has made.